src/gui/widgets/options_button.hpp

```diff
@@ -110,15 +122,27 @@ class options_button : public styled_widget
 	 */
 	state_t state_;
 
+//protected:
 	boost::container::stable_vector<menu_item> values_;
 
-protected:
+	/* toggle_states_ is used only to transfer checkbox values to and from the widget, and may
+	 * not be up to date at any other time.  Use values_ for all other operations */
+    boost::dynamic_bitset<> toggle_states_;
+
+	dialogs::drop_down_menu* droplist_;
+
+    void update_config_from_toggle_states();
+
 	unsigned selected_;
 
+	/* For widgets that update label depending on their value(s) */
+	virtual void update_label() {}
+
 private:
 	bool keep_open_;
 
 	/* Whether or not the item selected should be remembered if menu is re-opened */
+	/* Does not apply to multimenu_button */
```

Comment:
Why does it not apply to multimenu_button? I don't think there's any fundamental reason why it can't apply. Keep in mind that "selected" here (I'm pretty sure) means "highlighted", not "checkbox is active".

Reply:
selected really only applies to menu_button atm.  mmb does retain its state, but the variable in question is not relevant to that.

Reply:
Hmm, and for a generic action button you probably wouldn't want to remember either (it should always open with the first item highlighted). So, maybe this should be moved into menu_button?

Reply:
It looks like it's going to only apply to mb, assuming there are no more buttons, but I put in in options because there are common functions (like left_click) that were almost exactly the same, but needed to act differently in the middle of the function and this was the way I could think of to avoid duplicating code.  If there's a better way...

Reply:
I'm not entirely sure what you mean by that… can you maybe describe the situation in a bit more detail?

Reply:
There are there functions that act differently if you're a menu button, but they only differ by just one line of code.  A couple of them are pretty easy to deal with because the extra line can be placed at the beginning or end of the function.  The problem child (signal_handler_left_button_click) looks like this:

Do some stuff
create DDM
if menubutton set DDM.start_selected (selected_index is "remembered") to true, else false
show DDM
Do some other stuff

The options I see would be:

1.  just make it the function virtual and override it in menu button.  This isn't so bad now, but there used to be a lot more code that would need to be replicated.
2. Just use the same function for everyone, but act differently if you're a menu button.  Here I've used a member variable that's different for the DDM, but I suppose looking at the instances type might be an option.
3. replace "if menubutton set..." with a virtual function, and have menu button override that (I do this elsewhere where menu button needs update_label() and no other button does
4. Some nonsense where you break the thing up into multiple functions to keep the common pieces the same, which I don't even think is worth addressing.

Reply:
Virtual functions is probably the way to solve this, generally speaking. But it depends on some of the details.

Your third option is one possibility – add an `init_drop_down_menu` function that can do whatever to the dropdown menu between it being constructed and shown. If you think there's any chance that a multimenu_button or any other hypothetical subclass might want to do something different here, then I'd say that's the way to go.

The persistent variable is not terrible if the intent is that it's a "protected" API – that is, it's up to the subclass to decide what to set it to. But if going with that option, this comment (that it doesn't apply to multimenu_button) is unnecessary. Each subclass can decide whether it applies.

I don't think option 4 is something to be dismissed out of hand. It does seem like overkill in this situation, but broadly speaking, having small functions with a single purpose is better for the protected/virtual interface, rather than big functions that do a whole bunch of things.

In conclusion: what you have now might be fine, so let's just leave it alone for now (but do update the comment).